File: ruby/test/test_short_checksum.rb
# frozen_string_literal: true

require "minitest/autorun"
require "baseh"

# Short checksum in expandable mode, spec section 22.
# Mirrors js/test/short-checksum.test.ts.
class TestShortChecksum < Minitest::Test
  TEST_KEY = "test-only-key-material-0001".b

  def expandable
    @expandable ||= Baseh::Baseh.new(Baseh.baseh_expandable_v1)
  end

  def raw(code)
    code.delete("-")
  end

  def assert_error(code)
    error = assert_raises(Baseh::BasehError) { yield }
    assert_equal code, error.code
  end

  # Find the first issuable id at or after from.
  def first_issuable(codec, from)
    from.upto(from + 10_000) do |id|
      codec.encode(id: id)
      return id
    rescue Baseh::BasehError
      next
    end
    raise "no issuable id from #{from}"
  end

  # Spec 22.5: the frozen tiers ship the feature on.
  def test_frozen_tier_shape
    profile = expandable.profile
    assert_equal 2, profile.checksum_length
    assert_equal 1, profile.short_checksum_length
    assert_equal 5, profile.short_checksum_until
    p = Baseh::Baseh.new(Baseh.baseh_expandable_p_v1(key_bytes: TEST_KEY, key_id: "test-01"))
    assert_equal 1, p.profile.short_checksum_length
    assert_equal 5, p.profile.short_checksum_until
  end

  def test_effective_checksum_length_per_generation
    profile = expandable.profile
    assert_equal 1, profile.effective_checksum_length(4)
    assert_equal 1, profile.effective_checksum_length(5)
    assert_equal 2, profile.effective_checksum_length(6)
    assert_equal 2, profile.effective_checksum_length(8)
  end

  # Spec 22.3: capacities follow the effective K.
  def test_generation_capacities_follow_the_effective_k
    assert_equal 39_304, expandable.generation_capacity(4) # 34^3
    assert_equal 1_336_336, expandable.generation_capacity(5) # 34^4
    assert_equal 1_336_336, expandable.generation_capacity(6) # one symbol buys the second checksum
    assert_equal 45_435_424, expandable.generation_capacity(7)
    assert_equal 1_544_804_416, expandable.generation_capacity(8)
  end

  def test_round_trips_first_and_last_issuable_id_of_generations_4_through_8
    (4..8).each do |l|
      first = first_issuable(expandable, expandable.generation_base(l))
      last = expandable.generation_base(l + 1) - 1
      [first, last].each do |id|
        begin
          code = expandable.encode(id: id)
        rescue Baseh::BasehError => e
          assert_equal "BLOCKED_CODE", e.code, "id #{id} blocked"
          next
        end
        assert_equal l, raw(code).length
        result = expandable.decode(code)
        assert_equal id, result.id
        assert_equal code, result.canonical_code
      end
    end
  end

  def test_short_normal_boundary
    last_short = expandable.generation_base(6) - 1 # 1,375,639
    first_normal = expandable.generation_base(6) # 1,375,640
    a = raw(expandable.encode(id: last_short))
    assert_equal 5, a.length
    assert_equal 4, a.length - 1 # 1 checksum symbol at length 5
    assert_equal last_short, expandable.decode(a).id
    b = raw(expandable.encode(id: first_normal))
    assert_equal 6, b.length
    assert_equal 4, b.length - 2 # 2 checksum symbols at length 6
    assert_equal first_normal, expandable.decode(b).id
  end

  def test_a_4_character_code_validates_against_exactly_1_checksum_symbol
    id = first_issuable(expandable, 0)
    code = raw(expandable.encode(id: id))
    assert_equal 4, code.length
    assert_equal code[3], Baseh::Checksum.calculate_checksum(expandable.profile, code[0...3], nil, 1)
    # Flipping the single checksum symbol fails.
    check = code[3]
    bad = check == "0" ? "1" : "0"
    assert_error("INVALID_CHECKSUM") { expandable.decode("#{code[0...3]}#{bad}") }
    # Appending a second checksum symbol changes the generation; the split
    # moves and the code fails (spec 19.7), it never validates as gen 4 + 2.
    assert_error("INVALID_CHECKSUM") { expandable.decode(code + check) }
  end

  def test_checksum_values_at_short_generations_use_modulus_35_not_1225
    id = first_issuable(expandable, 0)
    body = raw(expandable.encode(id: id))[0...3]
    short = Baseh::Checksum.calculate_checksum(expandable.profile, body, nil, 1)
    full = Baseh::Checksum.calculate_checksum(expandable.profile, body, nil, 2)
    assert_equal 1, short.length
    assert_equal 2, full.length
    assert_equal raw(expandable.encode(id: id))[3], short
  end

  # Spec 22.4: the separator threshold is still a function of total length.
  def test_separator_threshold_unchanged
    # Length 5 renders bare even though its body grew; length 6 splits.
    refute_includes expandable.encode(id: expandable.generation_base(5)), "-"
    assert_match(/\A...-...\z/, expandable.encode(id: first_issuable(expandable, expandable.generation_base(6))))
  end

  # Spec 22.4: the repetition scan covers body plus the short checksum.
  def test_repetition_scan_covers_the_short_checksum
    # Probe with the filter off to find an id whose 4-symbol raw code is a
    # run of 4 (necessarily spanning body and the single checksum symbol),
    # then confirm the frozen tier blocks it.
    probe = Baseh::Baseh.new(Baseh.baseh_expandable_v1.merge(max_repetition: 0))
    found = nil
    0.upto(expandable.generation_base(5) - 1) do |id|
      begin
        code = probe.encode(id: id)
      rescue Baseh::BasehError
        next
      end
      r = raw(code)
      next unless r.length == 4 && /(.)\1{3}/.match?(r)

      found = id
      break
    end
    refute_nil found, "expected a gen-4 code with a run of 4"
    assert_error("BLOCKED_CODE") { expandable.encode(id: found) }
  end

  # Spec 22.2 validation.
  def test_rejects_the_fields_in_fixed_mode
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(Baseh.baseh_medium_v1.merge(short_checksum_length: 1, short_checksum_until: 5))
    end
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(Baseh.baseh_medium_v1.merge(short_checksum_until: 5))
    end
  end

  def test_rejects_short_checksum_length_at_or_above_checksum_length
    base = Baseh.baseh_expandable_v1
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(base.merge(short_checksum_length: 2, short_checksum_until: 5))
    end
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(base.merge(short_checksum_length: 3, short_checksum_until: 5))
    end
  end

  def test_rejects_short_checksum_until_below_min_length
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(Baseh.baseh_expandable_v1.merge(short_checksum_length: 1, short_checksum_until: 3))
    end
  end

  def test_rejects_min_length_at_or_below_short_checksum_length
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(
        Baseh.baseh_expandable_v1.merge(min_length: 1, short_checksum_length: 1, short_checksum_until: 5)
      )
    end
  end

  def test_rejects_short_checksum_until_without_short_checksum_length
    plain = Baseh.baseh_expandable_v1.merge(short_checksum_length: 0, short_checksum_until: 0)
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(plain.merge(short_checksum_until: 5))
    end
  end

  def test_rejects_a_non_integer_short_checksum_length
    assert_error("INVALID_PROFILE") do
      Baseh::Baseh.new(Baseh.baseh_expandable_v1.merge(short_checksum_length: 1.5, short_checksum_until: 5))
    end
  end

  def test_zero_or_absent_turns_the_feature_off
    off = Baseh::Baseh.new(
      Baseh.baseh_expandable_v1.merge(short_checksum_length: 0, short_checksum_until: 0)
    )
    assert_equal 0, off.profile.short_checksum_length
    assert_equal 1_156, off.generation_capacity(4)
    assert_equal 2, off.profile.effective_checksum_length(4)
    code = off.encode(id: 1_155)
    assert_equal 4, raw(code).length
    assert_equal 1_155, off.decode(code).id
  end

  def test_a_custom_short_checksum_window_round_trips_at_every_generation
    h = Baseh::Baseh.new(
      Baseh.baseh_expandable_v1.merge(
        profile_id: "short-window-test",
        min_length: 4,
        checksum_length: 2,
        short_checksum_length: 1,
        short_checksum_until: 6,
        permutation: { enabled: false },
        profanity: { mode: "none" },
        max_repetition: 0
      )
    )
    # Body sizes: 3, 4, 5 through length 6 (K = 1), then L - 2.
    assert_equal 34**3, h.generation_capacity(4)
    assert_equal 34**5, h.generation_capacity(6)
    assert_equal 34**5, h.generation_capacity(7) # K = 2 kicks in
    assert_operator h.generation_capacity(6), :>, h.generation_capacity(5)
    (4..8).each do |l|
      id = h.generation_base(l) + 7
      code = h.encode(id: id)
      assert_equal l, raw(code).length
      assert_equal id, h.decode(code).id
    end
  end
end

File: python/tests/test_short_checksum.py
"""Short checksum tests, mirroring js/test/short-checksum.test.ts (spec 22):
frozen tier shape, generation round trips and the short/normal boundary,
effective-K decode, validation errors, interactions with the repetition
filter and separator threshold, and a custom short-checksum window."""

import re
import unittest

from baseh import (
    Baseh,
    BasehError,
    INVALID_PROFILE,
    baseh_expandable_p_v1,
    baseh_expandable_v1,
    baseh_medium_v1,
    effective_checksum_length,
    generation_base,
    generation_capacity,
)
from baseh.checksum import calculate_checksum

_TEST_KEY = b"test-only-key-material-0001"


def _raw(code):
    return code.replace("-", "")


def _expect_error(fn, code):
    try:
        fn()
    except BasehError as err:
        if err.code != code:
            raise AssertionError(f"expected {code}, got {err.code}")
        return
    raise AssertionError(f"expected {code}")


def _first_issuable(codec, from_id):
    """Find the first issuable id at or after from_id."""
    for id in range(from_id, from_id + 10000):
        try:
            codec.encode(id)
            return id
        except BasehError:
            continue
    raise AssertionError(f"no issuable id from {from_id}")


class TestFrozenTierShape(unittest.TestCase):
    def setUp(self):
        self.codec = Baseh(baseh_expandable_v1())

    def test_ships_feature_on(self):
        profile = self.codec.profile
        self.assertEqual(profile.checksum_length, 2)
        self.assertEqual(profile.short_checksum_length, 1)
        self.assertEqual(profile.short_checksum_until, 5)
        keyed = Baseh(baseh_expandable_p_v1(_TEST_KEY, key_id="test-01"))
        self.assertEqual(keyed.profile.short_checksum_length, 1)
        self.assertEqual(keyed.profile.short_checksum_until, 5)

    def test_effective_checksum_length_per_generation(self):
        profile = self.codec.profile
        self.assertEqual(effective_checksum_length(profile, 4), 1)
        self.assertEqual(effective_checksum_length(profile, 5), 1)
        self.assertEqual(effective_checksum_length(profile, 6), 2)
        self.assertEqual(effective_checksum_length(profile, 8), 2)

    def test_generation_capacities_follow_effective_k(self):
        profile = self.codec.profile
        self.assertEqual(generation_capacity(profile, 4), 19683)  # 27^3
        self.assertEqual(generation_capacity(profile, 5), 531441)  # 27^4
        # one symbol buys the second checksum
        self.assertEqual(generation_capacity(profile, 6), 531441)
        self.assertEqual(generation_capacity(profile, 7), 14348907)
        self.assertEqual(generation_capacity(profile, 8), 387420489)


class TestRoundTripsAndBoundaries(unittest.TestCase):
    def setUp(self):
        self.codec = Baseh(baseh_expandable_v1())

    def test_first_and_last_issuable_of_generations_4_to_8(self):
        for length in range(4, 9):
            first = _first_issuable(
                self.codec, generation_base(self.codec.profile, length)
            )
            last = generation_base(self.codec.profile, length + 1) - 1
            for id in (first, last):
                with self.subTest(id=id):
                    try:
                        code = self.codec.encode(id)
                    except BasehError as err:
                        self.assertEqual(err.code, "BLOCKED_CODE")
                        continue
                    self.assertEqual(len(_raw(code)), length)
                    result = self.codec.decode(code)
                    self.assertEqual(result.id, id)
                    self.assertEqual(result.canonical_code, code)

    def test_short_normal_boundary(self):
        profile = self.codec.profile
        last_short = generation_base(profile, 6) - 1  # 551,123
        first_normal = generation_base(profile, 6)  # 551,124
        self.assertEqual(last_short, 551123)
        self.assertEqual(first_normal, 551124)
        a = _raw(self.codec.encode(last_short))
        self.assertEqual(len(a), 5)
        self.assertEqual(len(a) - 1, 4)  # 1 checksum symbol at length 5
        self.assertEqual(self.codec.decode(a).id, last_short)
        b = _raw(self.codec.encode(first_normal))
        self.assertEqual(len(b), 6)
        self.assertEqual(len(b) - 2, 4)  # 2 checksum symbols at length 6
        self.assertEqual(self.codec.decode(b).id, first_normal)

    def test_four_char_code_validates_against_one_checksum_symbol(self):
        profile = self.codec.profile
        id = _first_issuable(self.codec, 0)
        code = _raw(self.codec.encode(id))
        self.assertEqual(len(code), 4)
        self.assertEqual(
            code[3:], calculate_checksum(profile, code[:3], 1)
        )
        # Flipping the single checksum symbol fails.
        check = code[3]
        bad = "1" if check == "0" else "0"
        _expect_error(lambda: self.codec.decode(code[:3] + bad), "INVALID_CHECKSUM")
        # Appending a second checksum symbol changes the generation; the
        # split moves and the code fails (spec 19.7), it never validates as
        # gen 4 + 2.
        _expect_error(lambda: self.codec.decode(code + check), "INVALID_CHECKSUM")

    def test_short_generations_use_modulus_35(self):
        profile = self.codec.profile
        id = _first_issuable(self.codec, 0)
        body = _raw(self.codec.encode(id))[:3]
        short = calculate_checksum(profile, body, 1)
        full = calculate_checksum(profile, body, 2)
        self.assertEqual(len(short), 1)
        self.assertEqual(len(full), 2)
        self.assertEqual(_raw(self.codec.encode(id))[3:], short)

    def test_separator_threshold_is_total_length(self):
        # Length 5 renders bare even though its body grew; length 6 splits.
        self.assertNotIn(
            "-", self.codec.encode(generation_base(self.codec.profile, 5))
        )
        code = self.codec.encode(
            _first_issuable(self.codec, generation_base(self.codec.profile, 6))
        )
        self.assertRegex(code, r"^...-...$")

    def test_repetition_scan_covers_short_checksum(self):
        # A run of 4 that spans body and the single checksum symbol must be
        # blocked. The scan rule is profile-independent, so use a small
        # permutation-free profile where such a code is guaranteed and fast
        # to find, then confirm the filter blocks it.
        shape = {
            "profileId": "short-rep-test",
            "mode": "expandable",
            "bodyAlphabet": "AB",
            "minLength": 4,
            "checksumAlphabet": "0AB",
            "checksumLength": 2,
            "shortChecksumLength": 1,
            "shortChecksumUntil": 5,
            "caseSensitive": False,
            "separator": "",
            "separatorMinLength": 0,
            "grouping": [],
            "aliases": {},
            "permutation": {"enabled": False},
            "profanity": {"mode": "none"},
            "maxRepetition": 0,
        }
        probe = Baseh(shape)
        found = None
        for id in range(0, 2000):
            raw = _raw(probe.encode(id))
            if len(raw) >= 4 and re.search(r"(.)\1{3}$", raw):
                found = id
                break
        self.assertIsNotNone(found, "expected a code ending in a run of 4")
        blocked = Baseh({**shape, "maxRepetition": 4})
        _expect_error(lambda: blocked.encode(found), "BLOCKED_CODE")


class TestValidation(unittest.TestCase):
    def setUp(self):
        self.base = baseh_expandable_v1()

    def test_rejects_fields_in_fixed_mode(self):
        _expect_error(
            lambda: Baseh(
                {
                    **baseh_medium_v1(),
                    "shortChecksumLength": 1,
                    "shortChecksumUntil": 5,
                }
            ),
            INVALID_PROFILE,
        )
        _expect_error(
            lambda: Baseh({**baseh_medium_v1(), "shortChecksumUntil": 5}),
            INVALID_PROFILE,
        )

    def test_rejects_short_length_at_or_above_checksum_length(self):
        for short in (2, 3):
            with self.subTest(short=short):
                _expect_error(
                    lambda: Baseh(
                        {
                            **self.base,
                            "shortChecksumLength": short,
                            "shortChecksumUntil": 5,
                        }
                    ),
                    INVALID_PROFILE,
                )

    def test_rejects_until_below_min_length(self):
        _expect_error(
            lambda: Baseh(
                {**self.base, "shortChecksumLength": 1, "shortChecksumUntil": 3}
            ),
            INVALID_PROFILE,
        )

    def test_rejects_min_length_at_or_below_short_length(self):
        _expect_error(
            lambda: Baseh(
                {
                    **self.base,
                    "minLength": 1,
                    "shortChecksumLength": 1,
                    "shortChecksumUntil": 5,
                }
            ),
            INVALID_PROFILE,
        )

    def test_until_alone_is_a_legal_zero_checksum_window(self):
        # Spec 22 amendment: the window field is the switch, so until + absent
        # length (defaults to 0) is the zero-checksum window, not an error.
        codec = Baseh({**self.base, "shortChecksumLength": 0, "shortChecksumUntil": 5})
        self.assertEqual(codec.profile.short_checksum_length, 0)
        self.assertEqual(effective_checksum_length(codec.profile, 4), 0)

    def test_rejects_length_without_until(self):
        plain = {**self.base, "shortChecksumLength": 0, "shortChecksumUntil": 0}
        _expect_error(
            lambda: Baseh({**plain, "shortChecksumLength": 1}), INVALID_PROFILE
        )

    def test_rejects_until_above_8(self):
        _expect_error(
            lambda: Baseh(
                {**self.base, "shortChecksumLength": 1, "shortChecksumUntil": 9}
            ),
            INVALID_PROFILE,
        )

    def test_accepts_until_8(self):
        codec = Baseh({**self.base, "shortChecksumLength": 1, "shortChecksumUntil": 8})
        self.assertEqual(effective_checksum_length(codec.profile, 8), 1)
        self.assertEqual(effective_checksum_length(codec.profile, 9), 2)

    def test_rejects_non_integer_short_length(self):
        _expect_error(
            lambda: Baseh(
                {**self.base, "shortChecksumLength": 1.5, "shortChecksumUntil": 5}
            ),
            INVALID_PROFILE,
        )

    def test_zero_or_absent_turns_feature_off(self):
        off = Baseh(
            {**self.base, "shortChecksumLength": 0, "shortChecksumUntil": 0}
        )
        self.assertEqual(off.profile.short_checksum_length, 0)
        self.assertEqual(generation_capacity(off.profile, 4), 729)
        self.assertEqual(effective_checksum_length(off.profile, 4), 2)
        code = off.encode(100)
        self.assertEqual(len(_raw(code)), 4)
        self.assertEqual(off.decode(code).id, 100)

    def test_custom_window_round_trips_every_generation(self):
        profile = {
            **self.base,
            "profileId": "short-window-test",
            "minLength": 4,
            "checksumLength": 2,
            "shortChecksumLength": 1,
            "shortChecksumUntil": 6,
            "permutation": {"enabled": False},
            "profanity": {"mode": "none"},
            "maxRepetition": 0,
        }
        codec = Baseh(profile)
        # Body sizes: 3, 4, 5 through length 6 (K = 1), then L - 2.
        self.assertEqual(generation_capacity(codec.profile, 4), 27**3)
        self.assertEqual(generation_capacity(codec.profile, 6), 27**5)
        self.assertEqual(generation_capacity(codec.profile, 7), 27**5)  # K = 2
        self.assertGreater(
            generation_capacity(codec.profile, 6),
            generation_capacity(codec.profile, 5),
        )
        for length in range(4, 9):
            with self.subTest(length=length):
                id = generation_base(codec.profile, length) + 7
                code = codec.encode(id)
                self.assertEqual(len(_raw(code)), length)
                self.assertEqual(codec.decode(code).id, id)


class TestZeroChecksumWindow(unittest.TestCase):
    """Spec 22 amendment: a short_checksum_length of 0 inside a set window
    means no checksum symbols at those lengths — generations are all body."""

    def setUp(self):
        self.base = baseh_expandable_v1()
        self.zero_profile = {
            **self.base,
            "profileId": "short-zero-test",
            "minLength": 4,
            "checksumLength": 2,
            "shortChecksumLength": 0,
            "shortChecksumUntil": 5,
            "permutation": {"enabled": False},
            "profanity": {"mode": "none"},
            "maxRepetition": 0,
        }
        self.codec = Baseh(self.zero_profile)

    def test_effective_k_zero_inside_window(self):
        profile = self.codec.profile
        self.assertEqual(effective_checksum_length(profile, 4), 0)
        self.assertEqual(effective_checksum_length(profile, 5), 0)
        self.assertEqual(effective_checksum_length(profile, 6), 2)

    def test_window_generations_are_all_body(self):
        profile = self.codec.profile
        self.assertEqual(generation_capacity(profile, 4), 27**4)
        self.assertEqual(generation_capacity(profile, 5), 27**5)
        self.assertEqual(generation_capacity(profile, 6), 27**4)  # K = 2 above

    def test_round_trips_generations_4_to_6(self):
        profile = self.codec.profile
        for length in range(4, 7):
            for id in (
                generation_base(profile, length),
                generation_base(profile, length + 1) - 1,
            ):
                with self.subTest(length=length, id=id):
                    code = self.codec.encode(id)
                    self.assertEqual(len(_raw(code)), length)
                    result = self.codec.decode(code)
                    self.assertEqual(result.id, id)
                    self.assertEqual(result.canonical_code, code)

    def test_checksum_of_zero_symbols_is_empty_string(self):
        profile = self.codec.profile
        id = generation_base(profile, 4)
        code = _raw(self.codec.encode(id))
        self.assertEqual(len(code), 4)
        self.assertEqual(calculate_checksum(profile, code, 0), "")

    def test_typo_at_zero_checksum_generation_not_detected(self):
        # Documented trade-off (spec 22): there is no checksum to fail, so a
        # mistyped body symbol silently decodes to a different id.
        profile = self.codec.profile
        id = generation_base(profile, 4) + 1
        code = _raw(self.codec.encode(id))
        last = code[3]
        replacement = "2" if last == "1" else "1"
        typed = code[:3] + replacement
        result = self.codec.decode(typed)  # no error
        self.assertNotEqual(result.id, id)

    def test_correction_never_engages(self):
        # With no checksum there is nothing to correct against: any body
        # decodes as-is, exactly like a no-checksum fixed profile.
        profile = self.codec.profile
        id = generation_base(profile, 5) + 3
        code = _raw(self.codec.encode(id))
        result = self.codec.decode(
            code, try_correction=True, confusion_profile="heavy"
        )
        self.assertEqual(result.id, id)
        self.assertFalse(result.corrected)
        last = code[4]
        typed = code[:4] + ("2" if last == "1" else "1")
        result2 = self.codec.decode(
            typed, try_correction=True, confusion_profile="heavy"
        )
        self.assertNotEqual(result2.id, id)
        self.assertFalse(result2.corrected)

    def test_repetition_scan_covers_all_body_code(self):
        # Spec 22.4: at a zero-checksum generation the raw code is all body,
        # so the repetition scan covers the body only.
        filtered = Baseh({**self.zero_profile, "maxRepetition": 4})
        found = None
        for id in range(0, generation_capacity(self.codec.profile, 4)):
            if re.search(r"(.)\1{3}", _raw(self.codec.encode(id))):
                found = id
                break
        self.assertIsNotNone(found, "expected a gen-4 code with a run of 4")
        _expect_error(lambda: filtered.encode(found), "BLOCKED_CODE")


class TestUntil8WindowBoundary(unittest.TestCase):
    def setUp(self):
        self.codec = Baseh(
            {
                **baseh_expandable_v1(),
                "profileId": "short-until-8-test",
                "minLength": 4,
                "checksumLength": 2,
                "shortChecksumLength": 1,
                "shortChecksumUntil": 8,
                "permutation": {"enabled": False},
                "profanity": {"mode": "none"},
                "maxRepetition": 0,
            }
        )

    def test_generation_8_short_generation_9_full(self):
        profile = self.codec.profile
        id8 = generation_base(profile, 8) + 5
        c8 = _raw(self.codec.encode(id8))
        self.assertEqual(len(c8), 8)
        self.assertEqual(c8[7:], calculate_checksum(profile, c8[:7], 1))
        self.assertEqual(self.codec.decode(c8).id, id8)
        id9 = generation_base(profile, 9) + 5
        c9 = _raw(self.codec.encode(id9))
        self.assertEqual(len(c9), 9)
        self.assertEqual(c9[7:], calculate_checksum(profile, c9[:7], 2))
        self.assertEqual(self.codec.decode(c9).id, id9)


if __name__ == "__main__":
    unittest.main()
